src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -202,4 +212,110 @@ public static double calculateTPCMetrics(
+	 *         wifiscan entry" (i.e., a single entry with its {@code signal}
+	 *         attribute modified to be the aggregated signal value instead of the
+	 *         value in just the most recent entry for that (AP serial number,
+	 *         BSSID) tuple. The returned map will only contain APs which received
+	 *         at least one non-obsolete wifiscan entry from a BSS.
+	 */
+	public static Map<String, Map<String, WifiScanEntry>> getAggregatedWifiScans(Modeler.DataModel dataModel,
+			long obsoletionPeriodMs,
+			Aggregator<Double> agg) {
+		if (obsoletionPeriodMs < 0) {
+			throw new IllegalArgumentException("obsoletionPeriodMs must be non-negative.");
+		}
+		Map<String, Map<String, WifiScanEntry>> aggregatedWifiScans = new HashMap<>();
+		for (Map.Entry<String, List<List<WifiScanEntry>>> apToScansMapEntry : dataModel.latestWifiScans.entrySet()) {
+			String ap = apToScansMapEntry.getKey();
+			List<List<WifiScanEntry>> scans = apToScansMapEntry.getValue();
+			if (scans.isEmpty()) {
+				continue;
+			}
+			/*
+			 * Flatten the wifiscan entries and sort in reverse chronological order. Sorting
+			 * is done just in case the entries in the original list are not chronological
+			 * already - although they are inserted chronologically, perhaps latency,
+			 * synchronization, etc. could cause the actual unixTimeMs to be out-of-order.
+			 */
+			List<WifiScanEntry> mostRecentToOldestEntries = scans.stream().flatMap(List::stream)
+					.sorted((entry1, entry2) -> {
+						return -Long.compare(entry1.unixTimeMs, entry2.unixTimeMs);
+					}).collect(Collectors.toUnmodifiableList());
+			// Split mostRecentToOldest into separate lists for each BSSID
+			// These lists will be in reverse chronological order also
+			Map<String, List<WifiScanEntry>> bssidToEntriesMap = new HashMap<>();
+			for (WifiScanEntry entry : mostRecentToOldestEntries) {
+				if (entry.bssid == null) {
+					continue;
+				}
+				bssidToEntriesMap.computeIfAbsent(entry.bssid, bssid -> new ArrayList<>());
+				bssidToEntriesMap.get(entry.bssid).add(entry);
+			}
+
+			// calculate the aggregate for each bssid for this AP
+			for (Map.Entry<String, List<WifiScanEntry>> bssidToWifiScanEntriesMapEntry : bssidToEntriesMap.entrySet()) {
+				String bssid = bssidToWifiScanEntriesMapEntry.getKey();
+				List<WifiScanEntry> entries = bssidToWifiScanEntriesMapEntry.getValue();
+				WifiScanEntry mostRecentEntry = entries.get(0);
+				agg.reset();
+				for (WifiScanEntry entry : entries) {
+					if (mostRecentEntry.unixTimeMs - entry.unixTimeMs > obsoletionPeriodMs) {
+						// discard obsolete entries
+						break;
+					}
+					// mostRecentEntry should always match the first entry (they are the exact same
+					// object)
+					if (matchesForAggregation(mostRecentEntry, entry)) {
+						aggregatedWifiScans.computeIfAbsent(ap, k -> new HashMap<>());
+						aggregatedWifiScans.get(ap).putIfAbsent(bssid, new WifiScanEntry(entry));
```

Comment:
- same here re: `computeIfAbsent` -> `get` doing double-lookup
- instead of `putIfAbsent` you should just use `computeIfAbsent` again, which will avoid unnecessarily calling the `new WifiScanEntry()` constructor